Ble conf

paul-szczepanek-arm opened the pull request:
### Description

Allow users to configure the BLE feature to disable functionality not needed by the device to save memory.

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->
@pan- 

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing   /workflow.html#pull-request-types).
-->
No changes made to existing programs. By default all options are enabled which is the current state.

Commit:
feat(ble): Add configuration options for enabling/disabling BLE roles

Commit:
feat(ble): Add preprocessor check to determine enabled roles are valid

A number of BLE roles depend on each other, checking within the target
configuration file for a valid configuration is infeasible. Move the
validation to the preprocessor and fail to compile if some required
roles are not enabled.

Commit:
feat(ble): Guard out GattClient code when role is disabled

Commit:
feat(ble): Guard out GattServer code when role is disabled

Commit:
feat(ble): Guard out SecurityManager code when role is disabled

Commit:
feat(ble): Guard out Observer code when role is disabled

Commit:
feat(ble): Guard out the CordioBLE initialisation code

paul-szczepanek-arm committed:
add all the features and rename existing non-roles to features

paul-szczepanek-arm committed:
require at least one role observer or broadcaster

paul-szczepanek-arm committed:
move config to ble feature

paul-szczepanek-arm committed:
implied features

paul-szczepanek-arm committed:
add more feature toggles in cordio init

paul-szczepanek-arm committed:
require macros to be defined

paul-szczepanek-arm committed:
connectable is a feature not role

paul-szczepanek-arm committed:
added #if guards to cordio ble

paul-szczepanek-arm committed:
fix BLE_CONNECTABLE_FEATURE macro

paul-szczepanek-arm committed:
revert accidental changes

paul-szczepanek-arm committed:
whitelist feature

paul-szczepanek-arm committed:
guard class fields with ifs

paul-szczepanek-arm committed:
gatt client signing

paul-szczepanek-arm committed:
fixed conf file (tabs and order)

paul-szczepanek-arm committed:
update descriptions

paul-szczepanek-arm committed:
git review changes

paul-szczepanek-arm committed:
ifdef out security features

paul-szczepanek-arm committed:
merge failure fix

paul-szczepanek-arm committed:
ifdef source files based on features

paul-szczepanek-arm committed:
more gap ifdefs

paul-szczepanek-arm committed:
complete gap feature ifdef

paul-szczepanek-arm committed:
fix gap ifdefs

paul-szczepanek-arm committed:
ble ifdefs

paul-szczepanek-arm committed:
broadcaster role cordio pass

paul-szczepanek-arm committed:
cordio observer role pass

paul-szczepanek-arm committed:
cordio peripheral role pass

paul-szczepanek-arm committed:
extended doesn't depends on phy

paul-szczepanek-arm committed:
cordio extended peripheral pass

paul-szczepanek-arm committed:
correct dependencies for central and peripheral

paul-szczepanek-arm committed:
partial security pass

paul-szczepanek-arm committed:
cordio security pass

paul-szczepanek-arm committed:
privacy pass

paul-szczepanek-arm committed:
gatt server pass

mostly phy and security issues

paul-szczepanek-arm committed:
tab fix

pan- committed:
BLE - add static interface helper.

pan- committed:
BLE - replace some usages of Gap::Handle_t by ble::connection_handle_t

pan- committed:
BLE - devirtualize pal::AttClient

pan- committed:
BLE - devirtualize pal::AttClientToGattClientAdapter

pan- committed:
BLE - Devirtualize pal::ConnectionEventMonitor

The event handler has been extracted out of the monitor declaration.

pan- committed:
BLE - Devirtualize pal::Gap

The event handler has been taken out of Gap declaration and the instantiation must provide an implementation and the type that plays the event handler role.

pan- committed:
BLE - Devirtualize pal::GattClient

The event handler has been taken out of GattClient declaration and an instantiation requires the actual implementation and the type that handle events.

pan- committed:
BLE - Devirtualize pal::SecurityManager

The event handler has been extracted out of SecurityManager declaration and instantion of the interface requires the implementation and event handler type.

pan- committed:
BLE - Devirtualize pal::SigningEventMonitor

The event handler has been extracted out of SigningEventMonitor declaration and SigningEventMonitor instantion requires the implementation and event handler type.

pan- committed:
BLE - Devirtualization of ::Gap

The interface definition now lives in ::ble::interface::LegacyGap. Implementation must export the implementation type as ble::impl::LegacyGap.

pan- committed:
BLE - Devirtualize ::ble::Gap

The interface definition now lives in ::ble::interface::Gap.
The implementation must export the implementation types as ::ble::impl::Gap.

pan- committed:
BLE - Devirtualize ::GattClient

Interface definition now lives in ble::interface::GattClient. An implementation must export the implementation type in ::ble::impl::GattClient.

pan- committed:
BLE - Devirtualize ::GattServer.

The interface is defined in ::ble::interface::GattServer and an implementation must export the implementation type ::ble::impl::GattServer.

pan- committed:
BLE - Devirtualize ::SecurityManager

The interface now lives in ::ble::interface::SecurityManager. The implementation type is expectected to exported as ble ::ble::impl::SecurityManager by the implementation.

pan- committed:
BLE - Cleanup deprecation implementation bits in BLE.cpp

pan- committed:
BLE - cleanup includes.

pan- committed:
BLE - Devirtualize ble::generic::GenericGap

Expected types are similar to the type expected by the constructor:
- PalGap
- PalSecurityManager
- ConnectionEventMonitorEventHandler

Note that for the PalGap we expect a **template** of the form PalGap<EventHandler>

pan- committed:
BLE - Devirtualize ble::generic::GattClient

GenericGattClient is parametized by two types:
- The template of the PalGattClient
- The SigningMonitorEventHandler

Note that the PalGattClient template must be of the form PalGattClient<EventHandler>. The event handler being the GenericGattClient.

pan- committed:
BLE - Devirtualize ::ble::generic::GenericSecurityManager

Two parameters are expected by the template:
- TPalSecurityManager a template class of the security manager of the form TPalSecurityManager<EventHandler>. The GenericSecurityManager is the event handler.
- SigningMonitor a template in the form SigningMonitor<Handler>.

pan- committed:
BLE - Devirtualization of the Cordio port.

pan- committed:
BLE - Devirtualization of the Maxim port .

pan- committed:
BLE - Devirtualization of the legacy NRF51822 port.

pan- committed:
BLE - Devirtualization of the NRF51 port.

pan- committed:
BLE - Devirtualization of the NRF52 softdevice port.

pan- committed:
BLE - accept tpp extensions for header files.

pan- committed:
BLE - Fix qualification of abort and terminate.

donatieng commented on features/FEATURE_BLE/ble/BLETypes.h:
```diff
@@ -528,6 +528,7 @@ struct att_security_requirement_t : SafeEnum<att_security_requirement_t, uint8_t
          */
         NONE,
 
+#if BLE_FEATURE_SECURITY
```
I'd rather not ifdef enums - it will not help with debugging (enum values will differ based on compilation options), and has no code-size benefits

paul-szczepanek-arm replied:
These are to stop invalid/unhandled values. Otherwise the user might use an enum value which the code will not know what to do or more likely do something wrong. The correct solution is to grow the code by adding parameter checking for unhandled values which I didn't opt for in this instance due to time.

paul-szczepanek-arm replied:
also irrelevant in the light of the discussion this morning about limiting the PR to user api and init code

donatieng reviewed (changes requested):
Thanks @paul-szczepanek-arm, as discussed with @pan- offline we probably don't need so much granularity in the `#ifdefs`. Regarding Gap for instance, I suggest that in the future (not required for this PR) we split in in a couple of base classes that get derived from based on what's included:
* GapObserver
* GapScanner
* GapCentral
* GapPeripheral
* ...

What do you think?

pan- reviewed (changes requested):
Risks are high if we got this implementation In as many configuration combination needs to be tested manually. This is due to conditional compilation applied across all layers. 

The minimum we can do is to restrict conditional compilation to initialisation of the Cordio stack and event handling within the cordio stack. I believe it is safe to keep conditional directive on **user facing** function. 

To do that we need to remove conditional directive from function declaration in: 
* Generic layer 
* Pal layer 
* User facing event handler types
* Enum 

Note that it is fine to keep conditional directive within a function body.

New areas needs to be covered as well: 
- [`is_feature_supported`](https://github.com/ARMmbed/mbed-os/blob/31341392e4536e791b48e05b8e14dd029c3ba2ea/features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioPalGap.cpp#L26): Configuration must override controller capabilities. 
- [`gap_handler`](https://github.com/ARMmbed/mbed-os/blob/31341392e4536e791b48e05b8e14dd029c3ba2ea/features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioPalGap.cpp#L428) must take advantage of the configuration. 
- [`sm_handler`](https://github.com/ARMmbed/mbed-os/blob/31341392e4536e791b48e05b8e14dd029c3ba2ea/features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioPalSecurityManager.cpp#L542): The body of this function is the only thing that must be guard by `#if BLE_FEATURE_SECURITY` in this file. 
- Body of event handlers in [CordioGattServer.cpp](https://github.com/ARMmbed/mbed-os/blob/31341392e4536e791b48e05b8e14dd029c3ba2ea/features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioGattServer.cpp#L882-L1029) are the only part that must be guarded by `#if BLE_FEATURE_GATT_SERVER` in that file.

pan- committed:
Fix rebase mistakes.

pan- committed:
BLE - remove conditional directive from value types

pan- committed:
BLE - Remove conditionnal directive within services

pan- committed:
BLE - Remove conditional directive removing functions in pal.

pan- committed:
BLE - Cleanup conditional directives in gap/Gap/

- Remove directives from event handler type
- Expose privacy default configuration  unconditionally
- Fix conditional test for connect and cancelConnect.

pan- committed:
BLE - Cleanup conditional directives in Gap.

- Remove conditional directives for member variables.
- Remove conditional directive for processXXX functions that are exclusively used by implementation.
-  Remove conditional directives on types exported by the class.

pan- committed:
BLE - Remove conditionnal directive within GattClient::WriteOp_t

pan- committed:
BLE - Remove conditional compilation of FileSecurityDb

pan- committed:
BLE - Cleanup GenericGap conditional directive

- Within constructor
- Within member declaration
- Private function declaration and definition.

pan- committed:
BLE - Cleanup Generic GattClient conditional directives.

- Inherithance list
- private functions exposed
- members declaration.

pan- committed:
BLE - Cleanup GenericSecurityManager conditional directives

- Inheritance lisr
- Within constructor
- Within member declaration
- Private function declaration and definition.

pan- committed:
BLE - Cleanup CordioGattServer conditional directives.

pan- committed:
BLE - Fix conditional code within CordioPalGenericAccessService.

pan- committed:
BLE - Cleanup CordioPalSecurityManager conditional directives.

- Within constructor
- Within member declaration
- Private function declaration and definition.

pan- committed:
BLE - Remove conditional directive in construction of Cordio objects.

pan- committed:
BLE - Small reordering to reduce diff.

pan- committed:
BLE - Conditionally include address rotation at compile time

pan- committed:
BLE - Conditionally expose SecurityManager::enableSigning

pan- committed:
BLE - Disable peer csrk retrieval if signing disabled.

pan- committed:
BLE - Disabling SecurityManager connection handling when security disabled.

pan- committed:
BLE - Fix call to implementation.

pan- committed:
BLE - Conditionnally remove LESC and privacy event processing.

pan- committed:
BLE - Fix initialization of extended advertising modules.

pan- committed:
BLE - restrict AttConnRegister to GattServer.

pan- committed:
BLE - Conditionally handle gap events.

pan- committed:
BLE - Return supported feature based on compile time features.

pan- committed:
BLE - Improve FEATURE_PRIVACY conditions.

paul-szczepanek-arm commented on features/FEATURE_BLE/source/services/DFUService.cpp:
```diff
@@ -14,11 +14,6 @@
 
-#include "BLERoles.h"
-
-#if BLE_FEATURE_GATT_SERVER
```
why? does not compile now

pan- replied:
addressed

paul-szczepanek-arm commented on features/FEATURE_BLE/source/services/UARTService.cpp:
```diff
@@ -14,11 +14,6 @@
 
-#include "BLERoles.h"
-
-#if BLE_FEATURE_GATT_SERVER
```
does in fact need gattserver

paul-szczepanek-arm commented on features/FEATURE_BLE/source/services/URIBeaconConfigService.cpp:
```diff
@@ -14,11 +14,6 @@
 
-#include "BLERoles.h"
-
-#if BLE_FEATURE_GATT_SERVER
```
reinstate pls

pan- committed:
BLE - Exclude services if GattServer is not available.

pan- committed:
BLE - Fix exclusion of getMaxAdvertisingSetNumber and getMaxAdvertisingDataLength

pan- committed:
BLE - Fix conditional directive.

pan- committed:
BLE - Fix conditional directive.

pan- committed:
BLE - Filter user facing implementation based on configuration

paul-szczepanek-arm commented on features/FEATURE_BLE/ble/services/URIBeaconConfigService.h:
```diff
@@ -25,6 +25,8 @@
 #endif
 
+#if BLE_FEATURE_GATT_SERVER
+
```
still fails
```suggestion
#if BLE_ROLE_BROADCASTER
```

paul-szczepanek-arm committed:
add broadcaster ifdef and pull headers into ifdef

cmonr commented on features/FEATURE_BLE/source/gap/Gap.tpp:
```diff
@@ -0,0 +1,866 @@
+/* mbed Microcontroller Library
```
Ooohhh. This was deleted, renamed, then added.

I was wondering why it looked like the GH interface was being weird.

`git mv <file>{cpp,tpp}` appears to not have been done.

cmonr reviewed (approved):
This is a _really_ nice refactor!
